src/Payments.Mvc/Controllers/PaymentsController.cs

```diff
@@ -208,18 +208,31 @@ public async Task<ActionResult> AddCoupon(string id, string code)
                 return RedirectToAction("Pay", new { id });
             }
 
+            var teamCoupons = await _dbContext.Coupons.Where(c => c.Team.Slug == invoice.Team.Slug).ToListAsync();
+
             // find the coupon on same team
-            var coupon = await _dbContext.Coupons
-                .Where(c => c.Team.Slug == invoice.Team.Slug)
-                .FirstOrDefaultAsync(c => string.Equals(c.Code, code, StringComparison.OrdinalIgnoreCase));
+            var coupon = teamCoupons.FirstOrDefault(c => string.Equals(c.Code, code, StringComparison.OrdinalIgnoreCase));
 
             if (coupon == null)
             {
                 ErrorMessage = "Coupon code not found.";
                 return RedirectToAction("Pay", new { id });
             }
 
-            if (coupon.ExpiresAt.HasValue && coupon.ExpiresAt.Value < DateTime.UtcNow.ToPacificTime().Date)
+            bool expiredCoupon = coupon.ExpiresAt.HasValue && coupon.ExpiresAt.Value < DateTime.UtcNow.ToPacificTime().Date;
+
+            if (expiredCoupon)
+            {
+                //Check for other coupons
+                coupon = teamCoupons.OrderByDescending(a => a.Id).FirstOrDefault(a => a.ExpiresAt == null || a.ExpiresAt.Value >= DateTime.UtcNow.ToPacificTime().Date);
```

Comment:
Should the second search for an unexpired coupon also filter on Code like the prior search?

Also, should the first search include a sort like the second one?

Reply:
@sprucely I think it works ok as a final catch all.